File: plugins/plugin-browser/typescript/src/actions/select.ts
import type {
  Action,
  ActionResult,
  Content,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  State,
} from "@elizaos/core";
import { ServiceType } from "@elizaos/core";
import type { BrowserService } from "../services/browser-service.js";
import {
  ActionError,
  handleBrowserError,
  ServiceNotAvailableError,
  SessionError,
} from "../utils/errors.js";

export const browserSelectAction: Action = {
  name: "BROWSER_SELECT",
  similes: ["SELECT_OPTION", "CHOOSE", "PICK"],
  description: "Select an option from a dropdown on the webpage",

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const browserEnabled =
      runtime.getSetting("ENABLE_BROWSER") === "true" ||
      runtime.getSetting("BROWSER_ENABLED") === "true";

    if (!browserEnabled) {
      return false;
    }

    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      return false;
    }

    const text = message.content?.text?.toLowerCase() ?? "";
    return text.includes("select") || text.includes("choose") || text.includes("pick");
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult | undefined> => {
    const service = runtime.getService<BrowserService>(ServiceType.BROWSER);
    if (!service) {
      const error = new ServiceNotAvailableError();
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "select option"
      );
      return {
        text: "Browser service is not available",
        success: false,
        data: {
          actionName: "BROWSER_SELECT",
          error: "service_not_available",
        },
        values: {
          success: false,
          errorType: "service_not_available",
        },
      };
    }

    const session = await service.getOrCreateSession();
    if (!session) {
      const error = new SessionError("No active browser session");
      handleBrowserError(
        error,
        callback as (content: { text: string; error?: boolean }) => Promise<unknown>,
        "select option"
      );
      return {
        text: "No active browser session",
        success: false,
        data: {
          actionName: "BROWSER_SELECT",
          error: "no_session",
        },
        values: {
          success: false,
          errorType: "no_session",
        },
      };
    }

    const text = message.content?.text ?? "";
    const match = text.match(/["']([^"']+)["']/);
    const option = match ? match[1] : "";

    const dropdownMatch = text.match(/from (?:the )?(.+)$/i);
    const dropdown = dropdownMatch ? dropdownMatch[1] : "dropdown";

    if (!option) {
      throw new ActionError("select", dropdown, new Error("No option specified to select"));
    }

    const result = await service.getClient().select(session.id, option, dropdown);
    if (!result.success) {
      throw new ActionError("select", dropdown, new Error(result.error ?? "Select failed"));
    }

    const responseContent: Content = {
      text: `I've selected "${option}" from the ${dropdown}`,
      actions: ["BROWSER_SELECT"],
      source: message.content?.source ?? "action",
    };

    await callback?.(responseContent);

    return {
      text: responseContent.text ?? "",
      success: true,
      data: {
        actionName: "BROWSER_SELECT",
        option,
        dropdown,
        sessionId: session.id,
      },
      values: {
        success: true,
        option,
        dropdown,
      },
    };
  },

  examples: [
    [
      {
        name: "{{user}}",
        content: { text: 'Select "United States" from the country dropdown' },
      },
      {
        name: "{{agent}}",
        content: {
          text: 'I\'ve selected "United States" from the country dropdown.',
          actions: ["BROWSER_SELECT"],
        },
      },
    ],
  ],
};

File: plugins/plugin-browser/rust/src/actions/select.rs
use crate::services::BrowserService;
use crate::types::ActionResult;
use crate::utils::{action_error, parse_select_action, session_error};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::error;

pub const SELECT_ACTION_NAME: &str = "BROWSER_SELECT";
pub const SELECT_SIMILES: &[&str] = &["SELECT_OPTION", "CHOOSE", "PICK"];
pub const SELECT_DESCRIPTION: &str = "Select an option from a dropdown on the webpage";

pub async fn browser_select(service: Arc<BrowserService>, message: &str) -> ActionResult {
    let session = match service.get_or_create_session().await {
        Ok(s) => s,
        Err(e) => {
            let err = session_error(e);
            return ActionResult::failure(err.user_message);
        }
    };

    let (option, dropdown) = parse_select_action(message);

    if option.is_empty() {
        let err = action_error("select", &dropdown, Some("No option specified to select"));
        return ActionResult::failure(err.user_message);
    }

    let client = service.get_client();
    let result = client.select(&session.id, &option, &dropdown).await;

    match result {
        Ok(response) if response.success => {
            let mut data = HashMap::new();
            data.insert(
                "actionName".to_string(),
                serde_json::json!(SELECT_ACTION_NAME),
            );
            data.insert("option".to_string(), serde_json::json!(option));
            data.insert("dropdown".to_string(), serde_json::json!(dropdown));
            data.insert("sessionId".to_string(), serde_json::json!(session.id));

            ActionResult::success(data)
        }
        Ok(response) => {
            let err = action_error("select", &dropdown, response.error.as_deref());
            ActionResult::failure(err.user_message)
        }
        Err(e) => {
            error!("Error in BROWSER_SELECT action: {}", e);
            let err = action_error("select", &dropdown, Some(&e));
            ActionResult::failure(err.user_message)
        }
    }
}
